src/cpp/database/database.cpp

```diff
@@ -1352,6 +1352,71 @@ std::vector<std::pair<EntityId, EntityId>> Database::get_entities_by_name(
     return entities;
 }
 
+void Database::erase(
+        EntityId& domain_id)
+{
+    // Check that the given domain_id corresponds to a known monitor.
+    // Upper layer ensures that the monitor has been stopped.
+    // Upper layer also ensures that the monitor_id is valid and corresponds to a known domain.
+    assert(EntityKind::DOMAIN == get_entity_kind(domain_id));
+
+    // The mutex should be taken only once. get_entity_kind already locks.
+    std::unique_lock<std::shared_timed_mutex> lock(mutex_);
+
+    for (auto reader : datareaders_[domain_id])
+    {
+        // Unlink related locators
+        for (auto locator : reader.second.get()->locators)
+        {
+            locators_[locator.first].get()->data_readers.erase(reader.first);
+        }
+        // Clear entity
+        reader.second.get()->DDSEndpoint::clear();
+        reader.second.get()->clear();
+    }
+    // Clear datareaders map
+    datareaders_[domain_id].clear();
+
+    // Remove datawriters and unlink related locators
+    for (auto writer : datawriters_[domain_id])
+    {
+        // Unlink related locators
+        for (auto locator : writer.second.get()->locators)
+        {
+            locators_[locator.first].get()->data_writers.erase(writer.first);
+        }
+        // Clear entity
+        writer.second.get()->DDSEndpoint::clear();
+        writer.second.get()->clear();
+    }
+    // Clear datawriters map
+    datawriters_[domain_id].clear();
+
+    // Remove topics
+    for (auto topic : topics_[domain_id])
+    {
+        // Clear entity
+        topic.second.get()->clear();
+    }
+    // Clear topics map
+    topics_[domain_id].clear();
+
+    // Remove participants and unlink related process
+    for (auto participant : participants_[domain_id])
+    {
+        // Unlink related process
+        processes_[participant.second.get()->process.get()->id].get()->participants.erase(participant.first);
```

Comment:
What if the participant is not linked to any process?

Reply:
Good catch. I think I should add a test checking this case :facepalm:

Reply:
The regression test has been added in fd2487725036a37ab28dcb599c8bc4512e74e171

The implementation has been fixed in b525e9190608c5254df916184404b2c7942a6959